src/query_plan/fetch_dependency_graph.rs

```diff
@@ -292,18 +634,568 @@ impl DeferTracking {
+    let Some(conditions) = &child.conditions else {
+        return Err(FederationError::internal(format!(
+            "Key edge {edge:?} should have some conditions paths",
+        )));
+    };
+    let conditions_nodes = compute_nodes_for_tree(
+        dependency_graph,
+        conditions,
+        stack_item.node_id,
+        stack_item.node_path.clone(),
+        stack_item.defer_context.clone(),
+        &Default::default(),
+    )?;
+    created_nodes.extend(conditions_nodes.iter().copied());
+    let (source_id, dest_id) = tree.graph.edge_endpoints(edge_id)?;
+    let source = tree.graph.node_weight(source_id)?;
+    let dest = tree.graph.node_weight(dest_id)?;
+    let source_type: CompositeTypeDefinitionPosition = source.type_.clone().try_into()?;
+    let dest_type: CompositeTypeDefinitionPosition = dest.type_.clone().try_into()?;
+    let path_in_parent = &stack_item.node_path.path_in_node;
+    let updated_defer_context = stack_item.defer_context.after_subgraph_jump();
+    let new_node_id = dependency_graph.get_or_create_key_node(
+        &dest.source,
+        &stack_item.node_path.response_path,
+        &dest_type,
+        ParentRelation {
+            parent_node_id: stack_item.node_id,
+            path_in_parent: Some(path_in_parent),
+        },
+        &conditions_nodes,
+        updated_defer_context.active_defer_ref.as_ref(),
+    )?;
+    created_nodes.insert(new_node_id);
+    for condition_node in conditions_nodes {
+        // If `condition_node` parent is `node_id`,
+        // that is the same as `new_node_id` current parent,
+        // then we can infer the path of `new_node_id` into that condition node
+        // by looking at the paths of each to their common parent.
+        // But otherwise, we cannot have a proper "path in parent".
+        let mut path = None;
+        let mut iter = dependency_graph.parents_relations_of(condition_node);
+        if let (Some(condition_node_parent), None) = (iter.next(), iter.next()) {
+            // There is exactly one parent
+            if condition_node_parent.parent_node_id == stack_item.node_id {
+                if let Some(condition_path) = condition_node_parent.path_in_parent {
+                    path = condition_path.strip_prefix(path_in_parent).map(Arc::new)
+                }
+            }
+        }
+        drop(iter);
+        dependency_graph.add_parent(
+            new_node_id,
+            ParentRelation {
+                parent_node_id: condition_node,
+                path_in_parent: path.as_ref(),
+            },
+        )
+    }
+    let input_type = dependency_graph.type_for_fetch_inputs(source_type.type_name())?;
+    let mut input_selections = NormalizedSelectionSet::empty(
+        // TODO: should this be a subgraph schema?
+        dependency_graph.supergraph_schema.clone(),
+        input_type.clone(),
+    );
+    let Some(edge_conditions) = &edge.conditions else {
+        // TODO: Can this ever be `None`?
```

Comment:
I don't think so (also in general in JS code whenever there was a chance for error it would use some validation/explicit throws for nice error messages -> non-null assert here would result in pretty bad stack trace)